DB設計

wss4616 opened the pull request:
#what
DB設計のためReadmeを作成

#why
DB設計のため

ER図：https://gyazo.com/aca599f83a8becfe06d4605aa1013501

wss4616 committed:
DB設計

#what
DB設計のためReadmeを作成

#why
DB設計のため

ER図：https://gyazo.com/aca599f83a8becfe06d4605aa1013501

TakumaKawashima commented on README.md:
```diff
@@ -1,24 +1,101 @@
+| nickname	           |string	|null: false
+| email                |string	|null: false
+| encrypted_password	 |string	|null: false
+| user_image	         |string	|
```
```ruby
| user_image |string |
```
今回の実装ではユーザーのプロフィール画像は実装条件に含まれていませんので、こちらは削除しておきましょう。

TakumaKawashima commented on README.md:
```diff
@@ -1,24 +1,101 @@
-application up and running.
+Association
+has_many :products dependent: :destroy
+belongs_to :destination dependent: :destroy
```
```ruby
belongs_to :destination dependent: :destroy
```
住所情報を管理するテーブルは、実装条件では購入を管理するテーブルとだけ紐付いていますので、こちらは削除しておきましょう。

TakumaKawashima commented on README.md:
```diff
@@ -1,24 +1,101 @@
+| family_name	         |string	|null: false
+| first_name	         |string	|null: false
+| family_name_kana	   |string	|null: false
+| first_name_kane	     |string	|null: false
```
```ruby
| family_name	         |string	|null: false
| first_name	         |string	|null: false
| family_name_kana	   |string	|null: false
| first_name_kane	     |string	|null: false
```
ユーザー名などは24行目の外部キーカラムの情報から取得ができますので、こちらは削除しておきましょう。

TakumaKawashima commented on README.md:
```diff
@@ -1,24 +1,101 @@
+| city	               |string	|null: false
+| address	             |string	|null: false
+| building_name	       |string	|
+| phone_number	       |string	|
```
```ruby
| phone_number |string |
```
電話番号は必ず入力が必要になる項目ですので、入力されていなければ登録ができないようにしておきましょう。

TakumaKawashima commented on README.md:
```diff
@@ -1,24 +1,101 @@
+| family_name_kana	   |string	|null: false
+| first_name_kane	     |string	|null: false
+| |post_code	         |string	|null: false
+| prefecture	         |string	|null: false
```
```ruby
| prefecture |string |null: false
```
active_hashのカラム名は以下制約がありますので、修正しておきましょう。
```
・カラム名の末尾に「_id」を付与すること
・データ型は「integer型」を使用すること
```

TakumaKawashima commented on README.md:
```diff
@@ -1,24 +1,101 @@
 
-* System dependencies
+Association
+belongs_to :user
```
```ruby
belongs_to :user
```
購入を管理するテーブルとのアソシエーションんを組みましょう。

TakumaKawashima commented on README.md:
```diff
@@ -1,24 +1,101 @@
+###destinationテーブル###
+| Column               | Type   | Options                      |
+| -------------------- | ------ | -----------------------------|
+| user_id	             |integer	|null: false, foreign_key: true
```
```ruby
| user_id |integer |null: false, foreign_key: true
```
外部キーカラムを記載する場合は
```
・データ型は「references」
・カラム名の末尾に「_id」は不要
```
になりますので、修正しておきましょう。

TakumaKawashima commented on README.md:
```diff
@@ -1,24 +1,101 @@
+| -------------------- | ------ | -----------------------------|
+| user_id	             |integer	|null: false, foreign_key: true
+| customer_id	         |string	|null: false
+| card_id	             |string	|null: false
```
指摘①
```ruby
| user_id	             |integer	|null: false, foreign_key: true
| customer_id	         |string	|null: false
| card_id	             |string	|null: false
```
外部キーカラムを記載する場合は
```
・データ型は「references」
・カラム名の末尾に「_id」は不要
```
になりますので、修正しておきましょう。

指摘②
orderテーブルに保存される外部キーカラムは以下２つのみあることが正しいので、修正を行いましょう。
```
・ユーザー情報を管理するテーブル
・商品情報を管理するテーブル
```

TakumaKawashima commented on README.md:
```diff
@@ -1,24 +1,101 @@
 
-* Database initialization
+Association
+belongs_to :user
```
```
・ユーザー情報を管理するテーブル
・商品情報を管理するテーブル
・住所情報を管理するテーブル
```
外部キーカラムは２つですが、アソシエーションんは上記３つが必要になります。修正を行いましょう。

TakumaKawashima commented on README.md:
```diff
@@ -1,24 +1,101 @@
+| Column               | Type   | Options                      |
+| -------------------- | ------ | -----------------------------|
+|name	                 |string	|null: false
+|price                 |string	|null: false
```
```ruby
|price |string |null: false
```
商品価格のカラムのデータ型は「integer」にしておきましょう。

TakumaKawashima commented on README.md:
```diff
@@ -1,24 +1,101 @@
+| -------------------- | ------ | -----------------------------|
+|name	                 |string	|null: false
+|price                 |string	|null: false
+|description	         |string	|null: false
```
```ruby
|description |string |null: false
```
商品の説明用カラムのデータ型は文字数が多くなる傾向にあるので、保存上限の多い「text」型にしておきましょう。

TakumaKawashima commented on README.md:
```diff
@@ -1,24 +1,101 @@
+|judgment	             |string	
+|category_id	         |integer	|null: false, foreign_key: true
+|brand_id	             |integer	|null: false, foreign_key: true
+|shipping_id	         |integer	|null: false, foreign_key: true
```
```ruby
|status	               |string	|null: false
|size                  |string	|null: false
|shipping_cost	       |string	|null: false
|shipping_days	       |string	|null: false
|prefecture_id	       |string	|null: false
|judgment	             |string	
|category_id	         |integer	|null: false, foreign_key: true
|brand_id	             |integer	|null: false, foreign_key: true
|shipping_id	         |integer	|null: false, foreign_key: true
```
active_hashのカラムは以下３つが必要になります。不要なカラムがいくつかありますので、実装条件を見直し、修正を行いましょう。
```
カテゴリー・商品の状態・配送料負担・発送元地域・発送までの日数
```

TakumaKawashima commented on README.md:
```diff
@@ -1,24 +1,101 @@
+|category_id	         |integer	|null: false, foreign_key: true
+|brand_id	             |integer	|null: false, foreign_key: true
+|shipping_id	         |integer	|null: false, foreign_key: true
+|user_id             	 |integer	|null: false, foreign_key: true
```
```ruby
|user_id |integer |null: false, foreign_key: true
```
外部キーカラムを記載する場合は

・データ型は「references」
・カラム名の末尾に「_id」は不要
になりますので、修正しておきましょう。

TakumaKawashima commented on README.md:
```diff
@@ -1,24 +1,101 @@
+|user_id             	 |integer	|null: false, foreign_key: true
+
+ Association
+ belongs_to :user dependent: :destroy
```
```ruby
belongs_to :user dependent: :destroy
```
このままでは商品が削除された際にユーザー情報も削除されてしまいますので、dependent: :destroyは削除しておきましょう。

TakumaKawashima reviewed (commented):
レビュー依頼ありがとうございます。
コメントいたしましたので、ご確認よろしくお願いいたします。

wss4616 committed:
DB設計

#what
DB設計のためReadmeを作成

#why
DB設計のため

ER図：https://gyazo.com/7ebc5fe6ac47535abaf9e8b09f4c4f95
※テーブル4つにする為作り直しています。

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,83 @@
+| -------------------- | ------ | ------------|
+| nickname	           |string	|null: false
+| email                |string	|null: false, unique:true
+| password              |string  |null: false
```
```ruby
| password              |string  |null: false
```
上記のパスワードの情報を保存するカラムの設定ですが、カラム名を`encrypted_password`に変更しておきましょう！


理由：
deviseを使用する場合、新規登録の際に
passwordというキーに対して入力された値は
暗号化されてencrypted_passwordというカラムに保存されます。

そのため、このencrypted_passwordというカラムを
passwordカラムに変更してしまうと
エラーが発生してしまいます。

DB設計の段階からdeviseの仕様に合わせておくことで、
エラーを未然に防ぐことができます。

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,83 @@
-# README
+###userテーブル###
```
userテーブルには苗字名前等のカラム等は必要ですので、修正前に戻しておきましょう。

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,83 @@
+|family_name	          |string	    |null: false
+|first_name	            |string	    |null:false
+|family_name_kana	      |string	    |null:false
+|first_name_kana	      |string	    |null: false
```
```ruby
|family_name	          |string	    |null: false
|first_name	            |string	    |null:false
|family_name_kana	      |string	    |null:false
|first_name_kana	      |string	    |null: false
```

こちらは削除しましょう。

理由：
購入履歴に関するテーブルからuserの情報を取得できるためです。

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,83 @@
+|city	                  |string	    |null:false
+|house_number	          |integer	  |null: false
+|building_name	        |string	    |
+|phone_number	          |string	    |
```
```ruby
|phone_number	          |string	    |
```
null:falseを追加しましょう。

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,83 @@
+| status	              |integer    |null:false, default: 0
+| delivery_charge_bearer|integer    |null:false, default: 0
+| prefecture_id	        |integer    |null: false, default: 0
+| delivery_days	        |integer    |null:false, default: 0
```
```ruby
| brand	                |string	
| status	              |integer    |null:false, default: 0
| delivery_charge_bearer|integer    |null:false, default: 0
| prefecture_id	        |integer    |null: false, default: 0
| delivery_days	        |integer    |null:false, default: 0
```


商品の状態・配送料の負担などのカラムはactive_hashを用いての実装となります。

active_hashを用いたカラムの特徴は2点ございます：
①カラム名に`_id`が付与されること
active_hashを用いたカラムは、具体的な値ではなくid値（数値）を保存します。
この場合は以下のように、カラムに_idを付与しておくことで管理がしやすくなります。

```ruby
brand_id
```

②integer型を付与すること
id値（数値）を保存する際は、string型ではなく
integer型でカラムを作成すると管理しやすくなります。
string型からinteger型への修正をお願いいたします。

また、null:falseも追加をお願いいたします。

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,83 @@
+| price                 |integer    |null:false
+| seller_id	            |references	|null: false, foreign_key: true
+| buyer_id	            |references	|foreign_key: true
+| category_id	          |references	|null: false, foreign_key: true
```
references型はアソシエーションを組むテーブルのカラムを設定します。
userの外部キーのみで問題ございません。

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,83 @@
+Association
+belongs_to :user
+belongs_to :category
+has_many :item_images, dependent: :destroy
```
```ruby
belongs_to :category
has_many :item_images, dependent: :destroy
```
こちらは削除しましょう。

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,83 @@
-* Database initialization
 
-* How to run the test suite
+###item_imagesテーブル###
```
```ruby
item_imagesテーブル
```
こちらは削除しましょう。
画像に関するテーブルはActiveStorageにより自動的に作られるためです。

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,83 @@
 
-* ...
+
+###categoriesテーブル###
```
```ruby
categoriesテーブル
```
こちらは削除しましょう。

プルダウンを用いてカテゴリー選択をしますが、カテゴリーの内容はテーブルには保存しないためです。

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,83 @@
+belongs_to :user
+
+
+###cardテーブル###
```
```ruby
cardテーブル
```
こちらは購入履歴に関するテーブルでよろしかったですか？
購入履歴に関するテーブルでしたら、「誰が」「何を」購入したかの情報が入ります。
つまり、usersテーブルとitemsテーブルの外部キーを設定します。

wss4616 replied:
こちらクレジットカード機能のテーブルでした。
必要になった際に追記したいと思いますので削除しました。
改めてPurchase Recordsテーブル（購入記録）を作成しました。

OhyaYouki reviewed (commented):
レビュー依頼ありがとうございます。

修正点確認させていただきました。
追加でコメントいたしましたので、ご確認いただけますと幸いです。
今回テーブルの削除の修正をお願いしていおります。それに伴い今一度アソシエーションの見直しもお願いいたします。

修正後、再度レビュー依頼フォームをご提出ください。

ご対応のほど、よろしくお願いいたします。

wss4616 committed:
DB設計

#what
DB設計のためReadmeを作成

#why
DB設計のため
ER図GYazo:https://gyazo.com/5d0e78d1f4e669533e3cb38728964c47

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,61 @@
+|family_name	          |string	    |null: false
+|first_name	            |string	    |null:false
+|family_name_kana	      |string	    |null:false
+|first_name_kana	      |string	    |null: false
```
生年月日のカラムを追加しておきましょう。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,61 @@
+|family_name_kana	      |string	    |null:false
+|first_name_kana	      |string	    |null: false
+
+Association
```
１２行目
細かい点ですが、マークダウンを適用しておくと見やすくなります。
例）
```
### Association
```

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,61 @@
+Association
+has_one :user_address, dependent: :destroy
+has_many :items, dependent: :destroy, foreign_key: :items
+has_many :Purchase Records,dependent: :destroy, foreign_key: :Purchase Records
```
１３〜１５行目
アソシエーションの内容を見直し、修正しましょう。

[参考]
今回のアプリでは、ユーザーテーブルは、配送先情報とは直接アソシエーションを組む必要はございません。

アソシエーションの記述では、外部キーの記述（例.`foreign_key: :items`）は不要となりますので、削除しておきましょう。

アソシエーションのあとは、スネークケースで記述しておきましょう。
「:Purchase Records」はキャメルケースのような記述となります。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,61 @@
+| status_id	               |integer    |null:false, default: 0
+| delivery_charge_bearer_id|integer    |null:false, default: 0
+| prefecture_id	           |integer    |null: false, default: 0
+| delivery_days_id         |integer    |null:false, default: 0
```
２４〜２７行目
`default: 0`は削除しておきましょう。

今後ActiveHashを使用していくカラムですが、「0」は保存しない値となる可能性が高いため、デフォルト値の設定はなくしておきましょう。

CHURO-MAU replied:
２７行目
ActiveHashを利用するカラム名は「単数形_id」とする必要がございます。
現状複数形となっておりますので、修正しておきましょう。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,61 @@
+| prefecture_id	           |integer    |null: false, default: 0
+| delivery_days_id         |integer    |null:false, default: 0
+| price                    |integer    |null:false
+| user_id	                 |references	|null: false, foreign_key: true
```
２９行目
外部キーをreferences型で記述したときは、カラム名についている「_id」は不要となります。
修正しておきましょう。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,61 @@
+
+Association
+belongs_to :user
+has_many :Purchase Records,dependent: :destroy, foreign_key: :Purchase Records
```
３３行目
ユーザーテーブルのアソシエーションにて、コメントさせていただいた内容と同様になります。
ご確認いただき、こちらも修正しておきましょう。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,61 @@
+has_many :Purchase Records,dependent: :destroy, foreign_key: :Purchase Records
+
+
+###user_addressテーブル（発送先情報）###
```
３６行目
テーブルは複数形となりますので、修正しておきましょう。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,61 @@
+| Column                | Type      | Options     |
+| ----------------------| ----------| ------------|
+|post_number	          |string	    |null:false
+|prefecture_id	        |integer	  |null:false, default: 0
```
４０行目
こちらもActiveHashを使用するカラムとなりますので、`default: 0`は削除しておきましょう。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,61 @@
+|post_number	          |string	    |null:false
+|prefecture_id	        |integer	  |null:false, default: 0
+|city	                  |string	    |null:false
+|house_number	          |integer	  |null: false
```
４２行目
番地にはハイフンを含むため、カラムの型は文字列型としておきましょう。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,61 @@
+|house_number	          |integer	  |null: false
+|building_name	        |string	    |
+|phone_number	          |string	    |null:false
+|user_id	              |references |	null: false, foreign_key: true
```
４５行目
こちらも外部キーをreferences型で記述されているので、カラム名についている「_id」は不要となります。
修正しておきましょう。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,61 @@
+|user_id	              |references |	null: false, foreign_key: true
+
+Association
+belongs_to :user
```
[以前のコメント内容](https://github.com/wss4616/furima-37914/pull/13#discussion_r873149108)に対する修正箇所が見受けられませんでした。
ご確認いただき、修正のほどよろしくお願いいたします。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,61 @@
+| Column                | Type      | Options     |
+| ----------------------| ----------| ------------|
+|user_id	              |references	|null: false, foreign_key: true
+|item_id                |references	|null: false, foreign_key: true
```
５６、５７行目
こちらも外部キーをreferences型で記述されているので、カラム名についている「_id」は不要となります。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,61 @@
+|user_id	              |references	|null: false, foreign_key: true
+|item_id                |references	|null: false, foreign_key: true
+
+Association
```
こちらのテーブルには、配送先情報とのアソシエーションが必要になります。
追記してみましょう。

次回レビュー依頼時には、**最新のER図**もご提出をお願いいたします。
**最新のER図**は、プルリクエストに添付いただければと思います。

CHURO-MAU reviewed (commented):
レビュー依頼ありがとうございます！

修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いいたします。

データベース設計は、たいへん難しいフェーズですが、重要な設計となりますので頑張っていきましょう！

修正後は、再度レビュー依頼フォームのご提出をお願いいたします。

wss4616 committed:
DB設計

#what
DB設計のためReadmeを作成

#why
DB設計のため

ER図：https://gyazo.com/42cd10887d3b34b8885d34291d72f459

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,63 @@
+|first_name	           |string	    |null: false
+|family_name_kana	     |string	    |null: false
+|first_name_kana	     |string	    |null: false
+|birthday              |references  |null: false
```
１１行目
生年月日のカラムの型を変更しましょう。
こちらは日付を保存できるカラムの型に変更しておきましょう。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,63 @@
+
+###Association###
+has_many :items, dependent: :destroy, foreign_key: :items
+has_many :Purchase_Records,dependent: :destroy, foreign_key: :Purchase_Records
```
１４、１５行目
以下はふような記述となりますので、削除しておきましょう。
```
foreign_key: :items

foreign_key: :Purchase_Records
```

また、`:Purchase_Records`を修正しておきましょう。
単語の頭文字は小文字となります。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,63 @@
+
+###Association###
+belongs_to :user
+has_many :Purchase_Records,dependent: :destroy, foreign_key: :Purchase_Records
```
３３行目
こちらも同様に修正しておきましょう。
・`:Purchase_Records`の単語の頭文字を小文字に修正
・以下を削除
```
foreign_key: :Purchase_Records
```

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,63 @@
+|post_number	          |string	    |null:false
+|prefecture_id	        |integer	  |null:false
+|city	                  |string	    |null:false
+|housenumber	          |integer	  |null: false
```
４２行目
[こちら](https://github.com/wss4616/furima-37914/pull/13#discussion_r873377336)の修正が反映されておりませんでした。
ご確認いただき、修正のほどよろしくお願いいたします。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,63 @@
+###Association###
+belongs_to :user
+belongs_to :item
+has_one :Purchase_Record
```
５１行目
テーブルの関係性を見直しましょう。
今回は、配送先情報テーブルが子、購入履歴テーブルが親という関係になります。
そのような関係になるようにアソシエーションを修正してみましょう。

こちらも`:Purchase_Record`の単語の頭文字を小文字に直しておきましょう。

CHURO-MAU commented on README.md:
```diff
@@ -1,24 +1,63 @@
+###Association###
+belongs_to :user
+belongs_to :item
+belongs_to : user_address
```
６３行目
こちらも併せて関係背を見直し、アソシエーションを修正してみましょう。

今回、最新のER図が見受けられませんでした。
次回レビュー依頼時には、改めて最新のER図をご提出お願いいたします。

CHURO-MAU reviewed (commented):
レビュー依頼ありがとうございます！

修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いいたします。

修正後は、再度レビュー依頼フォームのご提出をお願いいたします。

wss4616 commented:
ER図がプルリクエストの際反映されていなかったようなのでコメントにも記載させていただきます。https://gyazo.com/351c5092db229327942aa9bcc4757e8b

wss4616 committed:
DB設計

ER図：https://gyazo.com/351c5092db229327942aa9bcc4757e8b

#what
DB設計のためReadmeを作成

#why
DB設計のため

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,63 @@
+###Association###
+belongs_to :user
+has_many :purchase_records,dependent: :destroy
+has_one :user_address
```
```ruby
has_one :user_address
```
こちらは削除しましょう。

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,63 @@
+|building_name	        |string	    |
+|phone_number	          |string	    |null:false
+|user	       　　        |references |	null: false, foreign_key: true
+|item                   |references |	null: false, foreign_key: true
```
```ruby
|user	       　　        |references |	null: false, foreign_key: true
|item                   |references |	null: false, foreign_key: true
```
こちらは削除しましょう。

理由：
購入記録テーブルに保存してある内容だからです。

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,63 @@
+|city	                  |string	    |null:false
+|house_number	          |string	  |null: false
+|building_name	        |string	    |
+|phone_number	          |string	    |null:false
```
購入履歴テーブルの外部キーカラムを追加しましょう。

OhyaYouki commented on README.md:
```diff
@@ -1,24 +1,63 @@
+###Association###
+belongs_to :user
+belongs_to :item
+belongs_to :purchase_record
```
```ruby
belongs_to :user
belongs_to :item
```
こちらは削除しましょう。

OhyaYouki reviewed (commented):
レビュー依頼ありがとうございます。

修正点確認させていただきました。
追加でコメントいたしましたので、ご確認いただけますと幸いです。
あともう少しで完成です！

修正後、再度レビュー依頼フォームをご提出ください。

ご対応のほど、よろしくお願いいたします。

wss4616 commented:
何度も何度もありがとうございます。
ER図：https://gyazo.com/6e0e6123bab523056cc18499fabfd98c

wss4616 committed:
DB設計

#what
DB設計のためReadmeを作成

#why
DB設計のため

ER図：https://gyazo.com/6e0e6123bab523056cc18499fabfd98c

OhyaYouki reviewed (commented):
レビュー依頼ありがとうございます！

修正点を確認いたしました。

**LGTM**とさせていただきます！
実装お疲れ様でした！

こちらのブランチは忘れずにMergeいただき、ローカルのMasterにPullした後に
次の実装のためのブランチを切って次の課題に取り組みましょう！

引き続き、開発頑張っていきましょう！応援しております。
